pkg/kv/kvserver/replica_destroy.go

```diff
@@ -132,7 +132,6 @@ func (r *Replica) postDestroyRaftMuLocked(ctx context.Context, ms enginepb.MVCCS
 	// Unhook the tenant rate limiter if we have one.
 	if r.tenantLimiter != nil {
 		r.store.tenantRateLimiters.Release(r.tenantLimiter)
-		r.tenantLimiter = nil
```

Comment:
Alternatively, we could protect `tenantLimiter` with `Replica.mu`, but I don't see much benefit. WDYT?

Reply:
Both options seem a little iffy. I don't want us to introduce another mutex acquisition in the critical path but at the same time it's not trivial to rearrange the users of this into an existing critical section.

Use-after-Release on the `tenantLimiter` might work - I looked at the code a bit and it seems like it doesn't destroy or pool any memory; it really just unlinks the limiter from the parent.

Let's go with the second option (the one you have chosen) but add a test in `tenantrate` that proves that you can use a limiter after it's been released and that this will not crash and also no longer influence the surrounding aggregating metric.

Reply:
I wrote a test. It does not panic after `Release`, but the parent metric is still updated. The `Destroy` method removes the child metric from the parent's tree, but still leaves the child->parent reference and keeps updating the parent metric. This API doesn't quite make sense to me, but it is what it is.

I think it's fine for counter metrics if some in-flight `Inc`rements slip through, but with the `Gauge` metrics we can't reliably reset it to 0 (and update the parent so that it no longer accounts for this value), a racing change may result in a permanent non-zero leak to the aggregate.

Reply:
I don't know, but the whole fact that `Send` can race with `Replica` destruction, and operate on a half-destroyed one, seems iffy.

Reply:
I think that the `Replica` "state machine" is confusing. Many things assume that there are only 2 states: uninitialized and initialized (hence the `IsInitialized` checks in many places, including `Send`, and assumption that many things synchronize through it), but in reality there is also `destroyStatus` which breaks this dichotomy.

Reply:
@RaduBerinde for this usage here

https://github.com/cockroachdb/cockroach/blob/e6e092fefc2a5ba0ff6342728f7b14f51e789899/pkg/ccl/multitenantccl/tenantcostserver/metrics.go#L170-L193

isn't it weird that these are gauges in the first place? And also, I don't see this getting released anywhere (we hold on to a map entry for a tenant forever), I assume this is intentional. So this should not be affected by not resetting gauges to zero but please let me know if we're missing something.